antarest/study/storage/study_upgrader/__init__.py

```diff
@@ -1,289 +1,72 @@
-    UpgradeMethod("710", "720", upgrade_720, []),
-    UpgradeMethod("720", "800", upgrade_800, [_GENERAL_DATA_PATH]),
-    UpgradeMethod("800", "810", upgrade_810, [_GENERAL_DATA_PATH, Path("input")]),
-    UpgradeMethod("810", "820", upgrade_820, [Path("input/links")]),
-    UpgradeMethod("820", "830", upgrade_830, [_GENERAL_DATA_PATH, Path("input/areas")]),
-    UpgradeMethod("830", "840", upgrade_840, [_GENERAL_DATA_PATH]),
-    UpgradeMethod("840", "850", upgrade_850, [_GENERAL_DATA_PATH]),
-    UpgradeMethod("850", "860", upgrade_860, [Path("input"), _GENERAL_DATA_PATH]),
-    UpgradeMethod("860", "870", upgrade_870, [Path("input/thermal"), Path("input/bindingconstraints")]),
-    UpgradeMethod("870", "880", upgrade_880, [Path("input/st-storage/clusters")]),
-]
+AVAILABLE_VERSIONS = ["700", "710", "720", "800", "810", "820", "830", "840", "850", "860", "870", "880"]
 
 
 class InvalidUpgrade(HTTPException):
     def __init__(self, message: str) -> None:
         super().__init__(HTTPStatus.UNPROCESSABLE_ENTITY, message)
 
 
-def find_next_version(from_version: str) -> str:
-    """
-    Find the next study version from the given version.
-
-    Args:
-        from_version: The current version as a string.
-
-    Returns:
-        The next version as a string.
-        If no next version was found, returns an empty string.
-    """
-    return next(
-        (meth.new for meth in UPGRADE_METHODS if from_version == meth.old),
-        "",
-    )
-
-
-def upgrade_study(study_path: Path, target_version: str) -> None:
-    with tempfile.TemporaryDirectory(suffix=".upgrade.tmp", prefix="~", dir=study_path.parent) as path:
-        tmp_dir = Path(path)
+class StudyUpgrader:
+    def __init__(self, study_path: Path, target_version: str):
         try:
-            src_version = get_current_version(study_path)
-            files_to_upgrade = can_upgrade_version(src_version, target_version)
-            files_to_retrieve = _copies_only_necessary_files(files_to_upgrade, study_path, tmp_dir)
-            _do_upgrade(tmp_dir, src_version, target_version)
-        except (StudyValidationError, InvalidUpgrade) as e:
-            logger.warning(str(e))
-            raise
-        except Exception as e:
-            logger.error(f"Unhandled exception : {e}", exc_info=True)
-            raise
+            version = StudyVersion.parse(target_version)
+        except ValueError as e:
+            raise InvalidUpgrade(str(e)) from e
         else:
-            _replace_safely_original_files(files_to_retrieve, study_path, tmp_dir)
-
+            self.app = UpgradeApp(study_path, version=version)
 
-def get_current_version(study_path: Path) -> str:
-    """
-    Get the current version of a study.
-
-    Args:
-        study_path: Path to the study.
+    def upgrade(self) -> None:
+        try:
+            self.app()
+        except Exception as e:
+            raise InvalidUpgrade(str(e)) from e
```

Comment:
We should not raise this type of exception, it seems to be meant for invalid input version, and reports a 422 HTTP status which is meant for client-side errors (invalid inputs).

Instead, we should probably let the exception through here.

Reply:
I agree, in the end it will be caught by the TaskJobService so I'll let the exception through